Intergrate SuperNova's `PublicParams`, benchmark NIVC, and Stablize Param API

winston-h-zhang opened the pull request:
Depends on https://github.com/lurk-lab/arecibo/pull/51 upstream. Resolves #681 and #682.

This PR addresses all of the issues surrounding public params. 

1. We integrate SuperNova's `PublicParams`, and extend the `Instance` API (more info below) to handle them smoothly. The cache will dynamically fetch/write/update params and sets of params depending on whether the user is asking for Nova or SuperNova proofs. 
2. We stabilize the param caching API by introducing a `Instance` structure, which uniquely represents exactly one `PublicParams`, `AuxParams`, or `CircuitParams`. `Instance`s compute a cache-key for the objects they represent and are themselves hashed to be put onto the file system by hash. For each instance, we also have a metadata json that contains useful information about it. Finally, we add a new CLI subcommand `lurk public-params` to manage this new cache structure.
3. Combining (1) and (2), we implement end to end NIVC benchmarks for Sha256.

porcuquine commented:
I don't think we should have per-lang directories. In the end, the `Lang` will include too much information to be fully-specified in a directory name; and anything but a flat namespace will only make it harder to find the needed params.

I think we should move toward  a model where parameter files have opaque names that serve as strong cache keys. That is, there should *never* be false cache hits: if the key you seek exists, then it has the params you require. This will require at least some degree of versioning of the underlying parameters — so that changes are reflected in the cache key.

I don't think we should try to be descriptive in the parameter names. If we want some degree of namespacing, it could be based on backend.

Since we do need to track the actual contents of the parameters (just not in their name strings, which are insufficient), I think we should add a metadata file with same name but different extension. That extension could be json, yaml, toml, lurk, whatever. (Probably json is fine for now.)

Then we can have some tooling that allows managing the cache, to allow for garbage collection of obsolete parameters — indexing, listing, moving (or at least identifying) specific params, etc.

huitseeker commented on src/public_parameters/disk_cache.rs:
```diff
@@ -34,36 +36,54 @@ where
 
-    pub(crate) fn set(&self, key: &str, data: &PublicParams<'static, F, C>) -> Result<(), Error> {
-        let file = File::create(self.key_path(key)).expect("failed to create file");
+    pub(crate) fn write_bytes(
```
Is this a write function?

huitseeker reviewed (dismissed):
## Top-level items
- figuring out if we need to hash to field elements, abstaining if not,
- figuring out if we need three `Kind`s, merging if not,
- using groups to avoid benchmark copy-pasta
- documenting the cache key computation practice this exemplifies properly (see below).

## Doc details

Overall, there are three behaviors, Supernova, IVC, and NIVC, that are tied to two `circuit_cache_key` functions. 

I understand (I think) why you're not setting up trait machinery to require every way of interfacing into a prover *provides* an implementation  that later function (our interfaces to Nova / SuperNova are due to fuse at some point), but I think we *must* document in plain English: 

- what kind of instances there are,
- how they dispatch to those cache key computations,
- what the inputs to this function are. In particular, I think we really want to document our choice of the `StepCircuit` iterated at rc=2 to capture the R1CSShape + its glue (IIUC),
- i.e. this practice is at the moment built-in to a couple of function definitions but to make them a practice, we need to explain what's going on (and have a link pointing to the doc of instance.rs on top of the circuit_cache_key functions),
- in order to help define what's going on, we can use new terms that are hitherto unused: "fingerprint" perhaps?

winston-h-zhang commented:
@huitseeker I just did a first pass of your comments. The original names were confusing, I've renamed the `Kind` enum arms and added some documentation. 

1. `compute_cache_key`: This was kept to return `F` because the cache keys (just their digest) I get for the circuits are `F`s, so I wanted to keep them consistent.
2. I realized that the entire `MemCache`/`DiskCache` structure is quickly becoming more of a hinderance than a useful abstraction. I think we could remove them and give the responsibility of managing files to `Instance`. I'm open to any other ideas too, if you have any suggestions.

huitseeker commented on benches/sha256.rs:
```diff
@@ -371,6 +365,166 @@ fn prove_compressed_benchmarks(c: &mut Criterion) {
+
+//             let frames = &prover
+//                 .get_evaluation_frames(ptr, env, store, limit, lang_rc.clone())
+//                 .unwrap();
+
+//             b.iter_batched(
+//                 || (frames, lang_rc.clone()),
+//                 |(frames, lang_rc)| {
+//                     let (proof, _, _, _, _) = prover.prove(&pp, frames, store, lang_rc).unwrap();
+//                     let compressed_result = proof.compress(&pp).unwrap();
+
+//                     let _ = black_box(compressed_result);
+//                 },
+//                 BatchSize::LargeInput,
+//             )
+//         },
+//     );
+// }
+
+// TODO: Add when compressed SNARK is implemented for SuperNova
+// fn prove_compressed_benchmarks(c: &mut Criterion) {
+//     tracing::debug!("{:?}", &*lurk::config::CONFIG);
+//     let reduction_counts = vec![10, 100];
+//     let batch_sizes = vec![1, 2, 5, 10, 20];
+//     let mut group: BenchmarkGroup<'_, _> = c.benchmark_group("prove_compressed");
+//     group.sampling_mode(SamplingMode::Flat); // This can take a *while*
+//     group.sample_size(10);
+//     let state = State::init_lurk_state().rccell();
+
+//     for &n in batch_sizes.iter() {
+//         for &reduction_count in reduction_counts.iter() {
+//             let prove_params = ProveParams {
+//                 arity: 1,
+//                 n,
+//                 reduction_count,
+//             };
+//             sha256_nivc_prove_compressed(prove_params, &mut group, state.clone());
+//         }
+//     }
+// }
```
What's the value of having this commented out given we have access to prove_compressed for IVC above?

winston-h-zhang replied:
True

huitseeker commented on benches/sha256.rs:
```diff
@@ -390,8 +547,11 @@ cfg_if::cfg_if! {
+                ivc_prove_benchmarks,
+                ivc_prove_compressed_benchmarks,
+                nivc_prove_benchmarks
+                // TODO: Add when compressed SNARK is implemented for SuperNova
```
Nit: This is more functions in the same group rather than two benchmarking groups. You can pass several groups to a criterion_main! macro. The two differ in how their statistics are aggregated in graphs for example.

huitseeker commented on src/proof/supernova.rs:
```diff
@@ -592,4 +587,46 @@ where
+/// public params to stay in sync with the changes.
+///
+/// Note: For now, we use ad-hoc circuit cache keys.
+/// See: https://zulip.lurk-lab.com/#narrow/stream/47-systems/topic/Public.20Parameters
```
Please point to the doc of Instance here.

winston-h-zhang replied:
Done

huitseeker reviewed (commented):
A rebase would be great!

winston-h-zhang commented on src/public_parameters/instance.rs:
```diff
@@ -0,0 +1,212 @@
+    },
+};
+
+const NUM_HASH_BITS: usize = 250;
```
I'll expose `NUM_HASH_BITS` as `pub` in arecibo

huitseeker commented:
I'm surprised to see 03d9ca9 and 3d2a4fa in this PR, which are part of #678, is it possible some of these changes slipped in via a diff error?

winston-h-zhang commented:
Yeah, I think I messed up somewhere in the middle... hold on a sec

huitseeker reviewed (commented):
This directionally LGTM in terms of functionality. Thanks for the doc on `Instance`, which really helps. One way that I think you could make this better is 
- giving a struct to the circuit cache keys, instead of `Vec<String>`, to future-proof the changes we may make there,
- implementing `SimpleDigestible` for this type,
- reusing the existing `DigestComputer` in `digest.rs` to replace `compute_digest`,

I've also left inline comments that should help with CI

winston-h-zhang commented on src/public_parameters/mod.rs:
```diff
@@ -60,60 +60,163 @@ where
+            Err(Error::CacheError("failed to decode bytes".into()))
+        }
+    })
+}
```
Good point. I think we can address this along with the planned `DiskCache`/`MemCache` refactors, so I'm also fine leaving it as is for now.

huitseeker reviewed (commented):
This LGTM  (I would stamp but I don't want to create a confusion with this line still in the PR:
https://github.com/lurk-lab/lurk-rs/pull/648/files#diff-2e9d962a08321605940b5a657135052fbcef87b5e360662bb527c96d9a615542R133 )

huitseeker reviewed (approved):
Thanks a bunch!
